graphstorm-processing/docs/source/developer/developer-guide.rst

```diff
@@ -0,0 +1,227 @@
+---------------
+
+The project is set up using ``poetry`` to make easier for developers to
+jump into the project.
+
+The steps we recommend are:
+
+Install JDK 8, 11 or 17
+~~~~~~~~~~~~~~~~~~~~~~~
+
+PySpark requires a compatible Java installation to run, so
+you will need to ensure your active JDK is using either
+Java 8 or 11.
+
+On MacOS you can do this using ``brew``:
+
+.. code-block:: bash
+
+    brew install openjdk@11
+
+On Linux it will depend on your distribution's package
+manager. For Ubuntu you can use:
+
+.. code-block:: bash
+
+    sudo apt install openjdk-11-jdk
+
+On Amazon Linux 2 you can use:
+
+.. code-block:: bash
+
+    sudo yum install java-11-amazon-corretto-headless
+    sudo yum install java-11-amazon-corretto-devel
+
+Install pyenv
+~~~~~~~~~~~~~
+
+``pyenv`` is a tool to manage multiple Python version installations. It
+can be installed through the installer below on a Linux machine:
+
+.. code-block:: bash
+
+   curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash
+
+or use ``brew`` on a Mac:
+
+.. code-block:: bash
+
+   brew update
+   brew install pyenv
+
+For more info on ``pyenv`` see https://github.com/pyenv/pyenv
+
+Create a Python 3.9 env and activate it.
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We use Python 3.9 in our images so this most closely resembles the
+execution environment on SageMaker.
+
+.. code-block:: bash
+
+   pyenv install 3.9
+   pyenv global 3.9
+
+..
+
+   Note: We recommend not mixing up conda and pyenv. When developing for
+   this project, simply ``conda deactivate`` until there's no ``conda``
+   env active (even ``base``) and just rely on pyenv+poetry to handle
+   dependencies.
+
+Install poetry
+~~~~~~~~~~~~~~
+
+``poetry`` is a dependency and build management system for Python. To install it
+use:
+
+.. code-block:: bash
+
+   curl -sSL https://install.python-poetry.org | python3 -
+
+Install dependencies through poetry
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now we are ready to install our dependencies through ``poetry``.
+
+We have split the project dependencies into the “main” dependencies that
+``poetry`` installs by default, and the ``dev`` dependency group that
+installs that dependencies that are only needed to develop the library.
+
+**On a POSIX system** (tested on Ubuntu, CentOS, MacOS) run:
+
+.. code-block:: bash
+
+   # Install all dependencies into local .venv
+   poetry install --with dev
+
+Once all dependencies are installed you should be able to run the unit
+tests for the project and continue with development using:
+
+.. code-block:: bash
+
+   poetry run pytest ./graphstorm-processing/tests
+
+You can also activate and use the virtual environment using:
+
+.. code-block:: bash
+
+   poetry shell
+   # We're now using the graphstorm-processing-py3.9 env so we can just run
+   pytest ./graphstorm-processing/tests
+
+To learn more about poetry see:
+https://python-poetry.org/docs/basic-usage/
+
+Use ``black`` to format code
```

Comment:
This is a choice to allow easier development of the library. `black` ensure all code is formatted in exactly the same style without any effort from the developer. Saves the developer from having to format code by hand to meet pylint criteria, since `black`-formatted code is also pylint-compatible.

Reply:
Sure, I'll mention this is optional for now, but I plan to introduce a black-formatting check for the library on Github, so it will become mandatory. To clarify the check will only apply to GSProcessing code.